---
title: "OCDC_Sumstats"
author: "Stephan"
date: "2024-08-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(data.table)
library(ggplot2)
install.packages("qqman")
library(qqman)

```

```{r}
# Paste in terminal 
# gsutil cp PASTE_OUTPUT_FROM_GWAS .
# bgzip -d log_reg.tsv.bgz
```

```{r}
data <- fread("log_reg.tsv", header = TRUE, sep = "\t")

# Split the locus column into CHR and BP
data[, c("CHR", "BP") := tstrsplit(locus, ":", type.convert = TRUE)]

# Convert CHR to numeric and handle non-numeric chromosomes
data[, CHR := gsub("chr", "", CHR)]
data[CHR == "X", CHR := 23]
data[CHR == "Y", CHR := 24]
data[CHR == "MT", CHR := 25]

# Convert CHR to numeric and check for NAs introduced by coercion
data[, CHR := as.numeric(CHR)]

# Ensure p-value column is numeric
data[, P := as.numeric(p_value)]

# Ensure P column contains finite values
data <- data[is.finite(data$P)]

# Display the cleaned data
head(data)
```

Manhattan Plot

```{r}
# Set the y-axis limit
max_y <- max(-log10(data$P), na.rm = TRUE)
ylim <- c(0, max(10, max_y))

manhattan(data, chr = "CHR", bp = "BP", snp = "alleles", p = "P", 
          main = "Manhattan Plot", col = c("blue4", "orange3"), ylim = ylim,
          suggestiveline = FALSE)

# Calculate Bonferroni significance threshold
bonferroni_threshold <- -log10(0.05 / nrow(data))

# Draw Bonferroni threshold line
abline(h = bonferroni_threshold, col = "green", lty = 1, lwd = 2)

# Add legend
legend("topright", legend=c("Bonferroni Threshold", "Standard GWAS Threshold"), 
       col=c("green", "red"), lty=1, lwd=2, bty="n")

```
QQ Plot 

```{r}
qq(data$P, main = "QQ Plot of P-values")

```

Prints the Top 50 results

```{r}
top_50 <- data[order(p_value)][1:50]
# Print the top 10 rows with the lowest p-values
print(top_50)
```

Writes top 50 results to a file

```{r}
write.csv(top_50, file = "top_50_lowest_p_values.csv", row.names = FALSE)

```

Prints results within a chromosome and P-value range

```{r}
filtered_data <- data[(data$CHR == 20 & data$P < 0.005)]
# Print the filtered data
print(filtered_data)
```

Inflation Factor 

```{r}
data[, chi_squared := qchisq(1 - P, df = 1)]

# Calculate the median of the observed chi-squared statistics
observed_median_chi2 <- median(data$chi_squared, na.rm = TRUE)

# Expected median chi-squared statistic under the null hypothesis
expected_median_chi2 <- qchisq(0.5, df = 1)

# Calculate the genomic inflation factor
lambda <- observed_median_chi2 / expected_median_chi2

# Print the genomic inflation factor
print(lambda)
```


